### ISSUE TITLE
Publicar ou descartar alterações em serviço

### ISSUE THREAD

**Issue by sahrosa** [servicosgovbr/editor-de-servicos#394]

**Comment by sahrosa** [comment]
Olá @almeidafab @mformiga @gustavo-freitas, vocês acham que a descrição acima está suficiente para o escopo do cartão? :)

**Comment by oliviagj** [comment]
@mformiga e @almeidafab trarão a segep para validar a linguagem

**Comment by gustavo-freitas** [comment]
Confirmar com Sá hipóteses quando usuário descarta alterações no serviço.

**Comment by gustavo-freitas** [comment]
Por mim, está bem completo @sahrosa. Caso usuário decida descartar alterações, passa a ser exibida será a última versão publicada, continuando na tela de edição. Faz sentido @sahrosa?

**Comment by sahrosa** [comment]
Faz sentido @gustavo-freitas. Acho que seria importante dar o feedback de que as alterações foram descartadas e que as edições previamente aprovadas estão sendo mostradas.

**Comment by kchnr** [comment]
@sahrosa, @gustavo-freitas em relação ao último item da lista, podemos fazer no próprio #479? Já que ainda não sabemos bem o tempo que demora a importação?

**Comment by sahrosa** [comment]
Acredito que sim, @kchnr. De acordo @gustavo-freitas ?

**Comment by gustavo-freitas** [comment]
Faz sentido @sahrosa. Adicionei informação no cartão #479.

**Comment by gustavo-freitas** [comment]
Conforme conversa com Fabrício, entendemos que:
- A possibilidade de 'Despublicar', 'Remover Alterações' e 'Excluir' adiciona ruído ao entendimento das funcionalidades do Editor pelo servidor. Nesse sentido, botão excluir na lista dos serviços, ficará oculto por agora. Confirmar comportamento com @mformiga.
- O botão 'Despublicar' não ficará ao lado do ícone de 'Remover Alterações', no topo da página do Editor. Decidimos mover esse botão para logo abaixo da lista de etapas, no lado esquerdo da edição do serviço. Além disso, seria ideal termos um link para a versão atual do serviço, indo diretamente para o Portal. Caso serviço não esteja publicado, botão deve estar desabilitado. Talvez, um texto alternativo mencionando que serviço não está no ar possa ser exibido também. @sahrosa faz sentido?

**Comment by oliviagj** [comment]
@sahrosa, @gustavo-freitas, @kchnr

Verificando a planilha de escopo, vi que tem uma pendência em relação à lista de serviços da página inicial do editor que é colocar alguma dica visual de que uma determinada página tem alterações com decisão de publicação/rejeição pendentes. isso faz parte deste cartão ou temos um outro para isso?

**Comment by sahrosa** [comment]
Acredito que não faça parte deste cartão @oliviagj. Não lembro se temos outro cartão pra isso, acho que o ideal seria criarmos um novo.

**Comment by gustavo-freitas** [comment]
Sim! Não achei nenhum cartão sobre essa indicação visual @sahrosa @oliviagj. Isso é prioritário @oliviagj ?

**Comment by gustavo-freitas** [comment]
Repassar os seguintes cartões antes de homologar:
- #314

**Comment by sahrosa** [comment]
@oliviagj @gustavo-freitas, @glauberramos e eu entendemos que "dica visual de que uma determinada página tem alterações com decisão de publicação/rejeição pendentes" é bastante importante e é um ajuste rápido. Cartão #426 criado!

**Comment by gustavo-freitas** [comment]
Beleza @sahrosa! Obrigado pela verificação. Outra coisa, achei os botões de 'Publicar' e 'Descartar Alterações' um pouco borrados. É assim, mesmo?

![image](https://cloud.githubusercontent.com/assets/2887992/11109211/5a7f7624-88d5-11e5-9234-ba08aab88f16.png)

**Comment by sahrosa** [comment]
Opa, aqui pra mim está ok. Vou verificar no monitor normal amanhã de novo!

**Comment by gustavo-freitas** [comment]
@sahrosa to vendo em alpha. Local não acontece isso. Talvez, tenha alguma relação com o brunch. Vou testar com mais calma amanhã.

**Comment by oliviagj** [comment]
Só para documentar [o ponto mais acima](https://github.com/servicosgovbr/editor-de-servicos/issues/394#issuecomment-153429764): a Segep, em reunião presencial no dia 5/11 com @mformiga, @almeidafab, eu e Yuli Hostenski, validou a linguagem proposta, conforme registrado em troca de emails.

**Comment by gustavo-freitas** [comment]
Cenários validados:

- Serviço existente editado e salvo. Remover alterações traz versão existente no portal 
- Serviço novo salvo. Todos os campos obrigatórios preenchidos. Publicar é realizado com sucesso.

Cenários problemáticos:

- Mesmo não preenchendo campos obrigatórios (palavras chave e legislação), serviço é publicado
- Quando serviço novo é salvo, browser fica sem exibir conteúdo por alguns segundos
- Quando campos obrigatórios de etapa não são preenchidos, validação indica 'Serviço ainda contém erros' mas campos problemáticos não são marcados em vermelho
- Serviço novo salvo. Serviço não publicado. Botão 'Descartar Alterações' não é desabilitado. Mensagem 'Não foi possível descartar as alterações' é exibida

**Comment by sahrosa** [comment]
Apenas reforçando, o item "Quando campos obrigatórios de etapa não são preenchidos, validação indica 'Serviço ainda contém erros' mas campos problemáticos não são marcados em vermelho" acontece comigo também.

**Comment by gustavo-freitas** [comment]
@sahrosa temos algum documento (além do XSD ) ou combinação que indica quais campos precisam ser obrigatórios?

**Comment by sahrosa** [comment]
Sim @gustavo-freitas, no subproduto IV.1 - Metadados de serviços!

**Comment by gustavo-freitas** [comment]
Cartão relacionado com #364.

**Comment by gustavo-freitas** [comment]
@kchnr 'Gratuidade' do serviço nunca é salva. Impedindo usuário de publicar serviço. Além disso, quando 'Gratuidade' é selecionada, campo custo não está sendo exibido para usuário.

**Comment by oliviagj** [comment]
>Além disso, quando 'Gratuidade' é selecionada, campo custo não está sendo
exibido para usuário.

É o comportamento esperado, @gustavo-freitas

**Comment by gustavo-freitas** [comment]
@oliviagj erro meu na hora de descrever o problema. Já corrigi o comentário. Campo custo não é exibido quando serviço tem custos.

**Comment by gustavo-freitas** [comment]
Na edição de um serviço, validações de campos estão acontecendo instantaneamente (logo depois que usuário abandona um determinado campo), não necessitando de nenhum tipo de operação como salvar ou publicar. Alguma sugestão aqui @sahrosa? Tu tens alguma sugestão?

**Comment by kchnr** [comment]
Gustavo freitas esse é o comportamento esperado, se o usuário passou por um
campo, será feito a validação desse campo.

2015-11-17 0:37 GMT-02:00 gustavo-freitas <notifications@github.com>:

**Comment by gustavo-freitas** [comment]
@kchnr tinha a impressão que antes só estávamos validando os campos obrigatórios depois de uma operação de salvar ou publicar. Coloquei a @sahrosa para saber a opinião dela, se ela acha essa validação quase instantânea interessante ou não. De qualquer forma, isso é um detalhe @kchnr.

**Comment by gustavo-freitas** [comment]
Comportamento observado até o momento:

Cenários                                      | Com Alterações (no serviço)     | Sem Alterações (no serviço)  |
----------------------------------------------|--------------------|-----------------|
Serviço: publicado                          | Botão Publicar (✓) Botão Desc (✓)    | Botão Publicar (✕) Botão Desc (✕)  |      
Serviço: despublicado                    | Botão Publicar (✓) Botão Desc (X)    | Botão Publicar (✕) Botão Desc (✕)  |

**Comment by gustavo-freitas** [comment]
Estou em dúvida sobre o seguinte comportamento: 

1. Novo serviço está sendo criado no editor
2. Nome utilizado no serviço é um nome existente (por exemplo: teste)
3. Usuário clica em salvar
4. Editor exibe mensagem de erro que serviço já existe, entretanto Editor continua salvando em segundo-plano. Todos os campos que não foram preenchidos são salvos na branch do serviço, sobrepondo todas as informações antigas do serviço. 

@kchnr esse é um comportamento esperado? Entendo que possa existir alguma relação com a funcionalidade de Importar, entretanto, usuário pode "perder" dados caso serviço não esteja publicado. Faz sentido?

**Comment by kchnr** [comment]
Maravilha Gustavo!

As validações que aparecem, são para a sua operação em específico. Então se
aparece a validação só no nome durante o Salvar, significa que esse campo
deve estar válido durante o Salvar. Se por algum motivo ele ignorou a
validação, isso é um problema.

Mas isso nada tem haver com este cartão, e este cartão pode ser homologado.
Mas o comportamento no caso do Salvar é:
1 - Validar se nome foi preenchido
2 - Validar se nome de serviço já existe

Se esses dois requisitos não estiverem certo, temos um problema na
validação de "Salvar Novo Serviço".



2015-11-17 2:30 GMT-02:00 gustavo-freitas <notifications@github.com>:

**Comment by gustavo-freitas** [comment]
@kchnr na minha perspectiva, temos um problema na validação do nome do serviço. O preenchimento do nome é checado, mas se usuário utiliza nome existente, Editor subscreve informações sem autorização do usuário. Acabei de reproduzir novamente. Como havíamos combinado de deixar essas operações de salvar/publicar/descartar 100%, levantei a bola aqui. Entretanto, concordo contigo. Vou abrir um defeito para essa validação de novo serviço.

**Comment by sahrosa** [comment]
Concordo com o @gustavo-freitas. Este é um problema bem sério, mas não vejo problema de ser um novo cartão!

**Comment by sahrosa** [comment]
Ops, enviei sem finalizar! Sobre a questão da validação dos campos, eu concordo que ela só deveria acontecer quando o usuário tenta publicar o serviço, até porque a premissa é salvar o serviço sem todas os campos estarem preenchidos corretamente necessariamente.

**Comment by gustavo-freitas** [comment]
Cartão validado com @mformiga e @almeidafab.

**Comment by gustavo-freitas** [comment]
Revisado com @almeidafab. Passando para produção.
